lib/vrl/compiler/src/vm/machine.rs

```diff
@@ -0,0 +1,533 @@
+                    // pass that as a hint to this OpCode so we only clone and fill up the stack when needed.
+                    match state.error.take() {
+                        Some(err) => {
+                            let err = Value::from(err.to_string());
+                            set_variable(ctx, variable, default.clone())?;
+                            set_variable(ctx, error, err.clone())?;
+                            state.push_stack(err);
+                        }
+                        None => {
+                            let value = state.pop_stack()?;
+                            set_variable(ctx, variable, value.clone())?;
+                            set_variable(ctx, error, Value::Null)?;
+                            state.push_stack(value);
+                        }
+                    }
+                }
+                OpCode::GetPath => {
+                    // Retrieves a value using the given path and pushes this onto the stack.
+                    let variable = state.next_primitive()?;
+                    let variable = &self.targets[variable];
+
+                    match &variable {
+                        Variable::External(path) => {
+                            let value = ctx.target().get(path)?.unwrap_or(Value::Null);
+                            state.stack.push(value);
+                        }
+                        Variable::Internal(ident, path) => {
+                            let value = match ctx.state().variable(ident) {
+                                Some(value) => match path {
+                                    Some(path) => {
+                                        value.get_by_path(path).cloned().unwrap_or(Value::Null)
+                                    }
+                                    None => value.clone(),
+                                },
+                                None => Value::Null,
+                            };
+
+                            state.stack.push(value);
+                        }
+                        Variable::None => state.stack.push(Value::Null),
+                        Variable::Stack(path) => {
+                            let value = state.pop_stack()?;
+                            let value = value.get_by_path(path).cloned().unwrap_or(Value::Null);
+                            state.stack.push(value);
+                        }
+                    }
+                }
+                OpCode::CreateArray => {
+                    // Creates an array from the values on the stack.
+                    // The next primitive on the stack is the number of fields in the array
+                    // followed by the values to be added to the array.
+                    let count = state.next_primitive()?;
+                    let mut arr = Vec::new();
+
+                    for _ in 0..count {
+                        arr.push(state.pop_stack()?);
+                    }
+                    arr.reverse();
```

Comment:
Unless this breaks something I haven't seen yet, I'd suggest reversing the order of items on the stack at compile time rather than runtime.

Reply:
Great idea!